Doc Refactor: Assessment Improvements (#1116)
fairlearn/fairlearn: docs/user_guide/assessment.rst

Assessment
==========

.. currentmodule:: fairlearn.metrics

In this section, we will describe the steps involved in performing a fairness

> assessment, and show how :class:`MetricFrame` can be

Author: Adjusted the text. LMK if that's what you meant.

nonbinary people, etc.). :footcite:cts:`crenshaw1991intersectionality`
offers a thorough background on the topic of intersectionality.

> One important thing to note here: we have assumed that every sensitive

Author: I had been wondering about that one. Adjusted.

feature is representable by a discrete variable.
This is not always the case: for example, the melanin content of a
person's skin (important for tasks such as facial recognition) will
not be taken from a small number of fixed values.
For now, features like this have to be binned.

> To reduce the risk of harms being blurred out by the binning,

Author: Let me simplify a little.

proxy variable measures the intended construct in a meaningful 
and useful way. It is important to ensure that the proxy is suitable 
for the social context of the problem you seek to solve. 
In particular, be careful of falling into one of the :ref:abstraction_traps.

The centerpiece of fairness assessment in Fairlearn are disaggregated metrics, 
which are metrics evaluated on slices of data. For example, to measure harms due to 
errors, we would begin by evaluating the errors on each slice of the data that 
corresponds to a group. If some of the groups are seeing much larger errors 
than other groups, we would flag this as a fairness harm.

Fairlearn provides the :class:`fairlearn.metrics.MetricFrame` class to help
with this quantification.
Suppose we have some 'true' values, some predictions from a model, and also

> a sensitive feature recorded for each:

Reviewer: ```suggestion
a sensitive feature recorded for each. The sensitive feature, denoted by `sf_data`, can take on one of four values. 
```

within more granular categories. In the case of demographic parity, we might 
need to review 
:math:`\E[h(X) \given A=a, D=d] = \E[h(X) \given D=d] \quad \forall a` 
where :math:`D` represents the feature(s) within :math:`X`` across which
members of the groups within :math:`A` are distributed. 
Demographic parity would then require that the prediction of the target  
variable is statistically independent of sensitive attributes conditional 
on D. Simply aggregating outcomes across high-level categories can be 
misleading when the data can be further disaggregated. 
It's important to review metrics across these more granular categories, 
if they exist, to verify that disparate outcomes persist across all levels 

> of aggregation. 

Reviewer: Suggest adding a link to the upcoming section on _Intersecting Groups_ during the discussion about considerations for group fairness. Something like "The section on :ref:`intersecting_groups` provides more detail on how to evaluate metrics across intersecting groups."

Author: Done.

comparisons problem, we recommend investigating `statistical techniques <https://www.statology.org/bonferroni-correction/>`_ 
meant to correct the errors produced by individual statistical tests. 

Fairlearn provides the :func:`demographic_parity_difference` and
:func:`demographic_parity_ratio` functions for computing demographic
parity measures for binary classification data, both of which return

> a scalar result

Reviewer: ```suggestion
a scalar result.
```

The first reports the absolute difference between the highest and
lowest selection rates :math:`a \in A` so a result of 0 indicates

> that demographic parity has been achieved (this does *not* automatically
> mean that the classifier is fair!).
> The second reports the ratio of the lowest and highest selection rates,
> so a result of 1 means there is demographic parity.

Reviewer: ```suggestion
that demographic parity has been achieved.
The second reports the ratio of the lowest and highest selection rates,
so a result of 1 means there is demographic parity. 
However, achieving demographic parity does *not*, automatically mean 
the classifier is fair!
```

Author: Agred.

comparisons problem, we recommend investigating `statistical techniques <https://www.statology.org/bonferroni-correction/>`_ 
meant to correct the errors produced by individual statistical tests. 

Fairlearn provides the :func:`demographic_parity_difference` and
:func:`demographic_parity_ratio` functions for computing demographic
parity measures for binary classification data, both of which return
a scalar result.
The first reports the absolute difference between the highest and
lowest selection rates :math:`a \in A` so a result of 0 indicates
that demographic parity has been achieved.
The second reports the ratio of the lowest and highest selection rates,
so a result of 1 means there is demographic parity.

> However, achieving demographic parity does *not* automatically mean

Reviewer 2: Not unique to demographic parity so maybe something like:
```suggestion
As with any fairness metric, achieving demographic parity does *not* automatically mean
```

contrast it with demographic parity. While demographic parity assesses the 
allocation of resources generally, equalized odds focuses on the allocation 
of resources that were actually distributed to 
members of that group (indicated by the positive target variable :math:`Y=1``). 
However, equalized odds makes the assumption 
that the target variable :math:`Y` is a good measurement of the phenomena 
being modeled, but that assumption may not hold if the measurement does not 
satisfy the requirements of construct validity.

Similar to the demographic parity case, Fairlearn provides
:func:`equalized_odds_difference` and :func:`equalized_odds_ratio`
to help with these calculations.
However, since equalized odds is based on both the true positive and
false positive rates, there is an extra step in order to return
a single scalar result.
For :func:`equalized_odds_difference`, we first calculate the
true positive rate difference and the true negative rate difference
separately.
We then return the larger of these two differences.

> *Mutatis mutandis*, :func:`equalized_odds_ratio` works similarly.

Author: I was following the principle of *Quidquid Latine dictum sit, altum viditur*.

capture the costs of missclassification disparities.

.. _scalar_metric_results:

> Scalar results from :code:`MetricFrame`

Author: Agreed - and this change in title makes me agree with @alliesaizan 's suggestion to have one of the page splits here (although that will be a separate PR).

This takes an underlying metric function, the name of the desired transformation, and
optionally a list of parameter names which should be treated as sample aligned parameters
(such as `sample_weight`).
The result is a function which builds the :class:`MetricFrame` internally and performs
the requested aggregation. For example:

.. doctest:: assessment_metrics
    :options:  +NORMALIZE_WHITESPACE

    >>> from fairlearn.metrics import make_derived_metric
    >>> recall_difference = make_derived_metric(metric=recall_score,
    ...                                        transform='difference')
    >>> recall_difference(y_true, y_pred,
    ...                   sensitive_features=sf_data)
    0.19999...
    >>> MetricFrame(metrics=recall_score,
    ...             y_true=y_true,
    ...             y_pred=y_pred,
    ...             sensitive_features=sf_data).difference()
    0.19999...

> We use :func:`fairlearn.metrics.make_derived_metric` to manufacture a number
> of such functions which will be commonly used:

Reviewer: ```suggestion
We use :func:`fairlearn.metrics.make_derived_metric` to manufacture a number
of such functions which are commonly used. The table below displays the aggregations 
that are compatible with each base metric:
```

Author: Like the addition sentence, but have tweaked slightly, since I don't think 'compatible' is quite the right word.

For example :code:`fairlearn.metrics.accuracy_score_difference` and
:code:`fairlearn.metrics.precision_score_group_min`.

Intersecting Groups
-------------------

The :class:`MetricFrame` class supports intersectional fairness in two ways:
multiple sensitive features, and control features.
Both of these can be used simultaneously.
One important point to bear in mind when performing an intersectional analysis
is that some of the intersections may have very few members (or even be empty).

> This can affect the confidence interval associated with the computed metrics.

Reviewer: Suggest adding a little more detail about how the confidence interval changes when intersectional groups are small or empty.

Author: Done.

    c     6       0.500000
          8       0.000000
    Name: recall_score, dtype: float64

If a particular intersection of the sensitive features had no members, then
the metric would be shown as :code:`NaN` for that intersection.
Multiple metrics can also be computed at the same time:

.. doctest:: assessment_metrics
    :options:  +NORMALIZE_WHITESPACE

    >>> from sklearn.metrics import precision_score
    >>> metric_2sf_multi = MetricFrame(
    ...     metrics={'precision':precision_score,
    ...              'recall':recall_score,
    ...              'count': count},
    ...     y_true=y_true,
    ...     y_pred=y_pred,
    ...     sensitive_features=s_f_frame
    ... )
    >>> metric_2sf_multi.overall
    precision     0.6
    recall        0.5
    count        18.0
    dtype: float64
    >>> metric_2sf_multi.by_group
               precision    recall  count
    SF 0 SF 1
    a    6      0.000000  0.000000    2.0
         8      0.500000  1.000000    2.0
    b    6      1.000000  0.666667    3.0
         8      1.000000  0.500000    3.0
    c    6      0.666667  0.500000    6.0
         8      0.000000  0.000000    2.0

Control Features
^^^^^^^^^^^^^^^^

> Control features (sometimes called 'conditional' features) enable more detailed

Reviewer: There's an example of when control features might be correlated with sensitive features, but it might also be helpful to provide a quick example of what a control feature might be in fairness analysis.

Author: Good point. I've taken a simpler solution, and moved the second paragraph up to complete this one, and had the remainder of this paragraph become the second paragraph.

    ... }
    >>> s_p = {
    ...     'recall_weighted' : { 'sample_weight':s_w },
    ...     'recall_weight_2' : { 'sample_weight':s_w_2 }
    ... }
    >>> weighted = MetricFrame(metrics=metrics,
    ...                        y_true=y_true,
    ...                        y_pred=y_pred,
    ...                        sensitive_features=pd.Series(sf_data, name='SF 0'),
    ...                        sample_params=s_p)
    >>> weighted.overall
    recall             0.500000
    recall_weighted    0.454545
    recall_weight_2    0.458333
    dtype: float64
    >>> weighted.by_group
          recall  recall_weighted  recall_weight_2
    SF 0
    a        0.5         0.500000         0.666667
    b        0.6         0.583333         0.600000
    c        0.4         0.250000         0.272727

More Complex Metrics
--------------------

So far, we have stuck to relatively simple cases, where the inputs are 1-D vectors of scalars,
and the metric functions return scalar values.
However, this need not be the case - we noted above that we were going to be vague as to the
contents of the input vectors and the return value of the metric function.

> Metric Function Returns Non-Scalar

Reviewer: ```suggestion
Non-Scalar Results from Metric Functions
```

Author: I like it.

    >>> metric_beta.overall
    0.56983...
    >>> metric_beta.by_group
    sensitive_feature_0
    a    0.365591
    b    0.850000
    c    0.468966
    Name: metric, dtype: float64

Per-Sample Arguments
^^^^^^^^^^^^^^^^^^^^

If there are per-sample arguments (such as sample weights), these can also be 
provided in a dictionary via the ``sample_params`` argument:

.. doctest:: assessment_metrics
    :options:  +NORMALIZE_WHITESPACE

    >>> s_w = [1, 2, 1, 3, 2, 3, 1, 2, 1, 2, 3, 1, 2, 3, 2, 3, 1, 1]
    >>> s_p = { 'sample_weight':s_w }
    >>> weighted = MetricFrame(metrics=recall_score,
    ...                        y_true=y_true,
    ...                        y_pred=y_pred,
    ...                        sensitive_features=pd.Series(sf_data, name='SF 0'),
    ...                        sample_params=s_p)
    >>> weighted.overall
    0.45...
    >>> weighted.by_group
    SF 0
    a    0.500000
    b    0.583333
    c    0.250000
    Name: recall_score, dtype: float64

If multiple metrics are being evaluated, then ``sample_params`` becomes a 
dictionary of dictionaries, with the first key corresponding matching that in 

> the dictionary holding the desired underlying metric functions.

Reviewer: I think this sentence is incomplete?

Author: That's a polite way of saying that I wrote gibberish.... ;-)

    >>> mf_conf_recall = MetricFrame(
    ...    metrics={ 'conf_mat':confusion_matrix, 'recall':recall_score },
    ...    y_true=y_true,
    ...    y_pred=y_pred,
    ...    sensitive_features=sf_data
    ... )
    >>> mf_conf_recall.overall
    conf_mat    [[2, 4], [6, 6]]
    recall                   0.5
    dtype: object
    >>> mf_conf_recall.by_group
                                 conf_mat  recall
    sensitive_feature_0
    a                    [[0, 2], [1, 1]]     0.5
    b                    [[1, 0], [2, 3]]     0.6
    c                    [[1, 2], [3, 2]]     0.4
    >>> mf_conf_recall.difference()
    conf_mat    None
    recall       0.2
    dtype: object

We see that the difference between group recall scores has been calculated, while a value of
:code:`None` has been returned for the meaningless 'maximum difference between two confusion matrices'
entry.

Inputs are Arrays of Objects
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

:class:`MetricFrame` can also handle cases when the :math:`Y_{true}` and/or :math:`Y_{pred}` vectors
are not vectors of scalars.
It is the metric function(s) which gives meaning to these values - :class:`MetricFrame` itself

> just slices the vectors up acorrding to the sensitive feature(s) and the control feature(s).

Reviewer: ```suggestion
just slices the vectors up according to the sensitive feature(s) and the control feature(s).
```

Author: Thanks for spotting this one!
